doc_indexer/src/indexing/indexer.py

```diff
@@ -83,47 +88,180 @@ def _load_documents(self) -> list[Document]:
+                    f"Indexed batch {i // CHUNKS_BATCH_SIZE + 1} of {len(chunks) // CHUNKS_BATCH_SIZE + 1}"
                 )
-
-                # Wait 3 seconds before processing next batch
-                if i + CHUNKS_BATCH_SIZE < len(all_chunks):
+                if i + CHUNKS_BATCH_SIZE < len(chunks):
                     time.sleep(3)
-
-            except Exception:
-                logger.exception(
-                    f"Error while storing documents batch {i//CHUNKS_BATCH_SIZE + 1} in HanaDB"
-                )
+            except Exception as e:
+                logger.error(f"Batch {i // CHUNKS_BATCH_SIZE + 1} failed: {e}")
                 raise
+        logger.info(
+            f"Successfully indexed {len(chunks)} markdown files chunks to {self.db.table_name} table."
+        )
+
+    def index(self) -> None:
+        """
+        Indexes markdown files in the specified directory by chunking them and storing the chunks in the database.
+
+        The process involves:
+            1. Loading and chunking markdown documents by specified headers.
+            2. Backing up the current main table (if it exists).
+            3. Deleting all data from the main table.
+            4. Storing new document chunks in the main table.
+            5. On error, restoring the main table from backup.
+
+        Raises:
+            Exception: If indexing fails and restoration from backup is attempted.
+        """
+
+        docs = self._load_documents()
+
+        # Chunk the documents by the headers.
+        chunks = create_chunks(docs, self.headers_to_split_on)
+
+        # Create a backup of the current main table (if exists).
+        logger.info(
+            f"Backing up current table {self.db.table_name} to {self.backup_table_name} if exists..."
+        )
+        self._copy_table(
+            self.db.table_name,
+            self.backup_table_name,
+            only_warn_if_table_inexistent=True,
+        )
+
+        try:
+            # Drop any old temporary table.
+            logger.info(f"Cleaning data from {self.db.table_name}.")
+            self.db.delete(filter={})
 
-        logger.info(f"Successfully indexed {len(all_chunks)} markdown files chunks.")
+            # Store all new document chunks in the temporary table.
+            logger.info(
+                f"Indexing {len(chunks)} markdown files chunks for {self.db.table_name}..."
+            )
+            self._index_chunks_in_batches(chunks)
+        except Exception:
+            logger.exception(
+                "Error during indexing. Attempting to restore from backup."
+            )
+            self._drop_table(self.db.table_name, only_warn_if_table_inexistent=True)
+            self._rename_table(
+                self.backup_table_name,
+                self.db.table_name,
+                only_warn_if_table_inexistent=True,
+            )
+            raise
```

Comment:
you can delete backup table if it exists with finally block

Reply:
Valid point.